### DISCUSSION TITLE
`corrurent` tests that use `mockImplementationOnce`?

### DISCUSSION THREAD

**Discussion by ValentinGurkov** [vitest-dev/vitest#4912] [Q&A] [answered]
Hello,

I want to understand the concepts a bit more if somebody can explain.
I run two tests under `describe.concurrent`. - test A and B.
Test B has `mockImplementationOnce` on a mocked function.
However, I imagine due to the concurrency, test A ends up using the mocked implementation instead of test B where it is defined.
If I remove the concurrency, the tests will execute as expected.
Do you know if this is expected? Should I opt for not using concurrency when I am mocking implementations inside a test?

Thanks!

**Comment by sheremet-va** [comment]
Since they run at the same time, it's possible that you are overriding the implementation. This is how it's running:

```ts
const spy = vi.fn()
await Promise.all(
  async () => spy.mockReturnedOnce(1),
  async () => spy.mockImplementationOnce(() => 2),
)
```

**Reply by ValentinGurkov** [reply]
Thanks, can we take a look at this example?

```ts
describe.concurrent('create', () => {
    it('should create a product successfully and publish Pub/Sub message', async ({ expect }) => {
        const result = await productService.create(membershipProductPayload);
        expect(result).toBeInstanceOf(MasterProduct);
        expect(pubSubApp.publish).toHaveBeenCalled();
    });

    it('should throw an error on database errors', async ({ expect }) => {
        const driverError = new DatabaseError('Some message', 5, 'error');
        driverError.constraint = ValidationConstraintEnum.UQ_SKU;
        const error = new QueryFailedError('test', [], driverError);
        masterProductRepositoryMock.save.mockImplementationOnce(() => {
            throw error;
        }); // masterProductRepositoryMock.save has implementation via vi.fn() in the upper scope
        await expect(productService.create(membershipProductPayload)).rejects.toThrow(....);
    });
});
```
It this example, `masterProductRepositoryMock.save.mockImplementationOnce` from the second test get's picked up on the first test when I set `describe.concurrent`

Do I need to wrap the whole test in an async function like in your example?

**Reply by sheremet-va** [reply, marked as answer]
No, I just explained how it's executed and why it doesn't work for you. It's a race condition in your code because you rely on a global variable. You cannot run these tests in parallel and expect it to work.

**Reply by ValentinGurkov** [reply]
Thanks, I'll need to learn more in regards of how to change the implementation of a mocked library in concurrent tests
